File: Prácticas/01-Constante Madelung/Practica1V2.py
# Versión 2 del cálculo de la constante de madelung tratando de optimizar el código al máximo.

from math import sqrt, fabs
from timeit import default_timer
from itertools import product   
import numpy as np

def constante_madelung(L: float) -> float:
    
    # Primero inicializamos en cero la constante.
    M = 0

    # Creamos todas las posibles coordenadas con la función product de la librería itertools
    bucle = list(product(np.arange(-L, L+1),repeat=3))
    bucle.remove((0,0,0)) # Evitamos dividir por cero quitando la coordenada (0,0,0)

    for coordenada in bucle:
        i,j,k = coordenada[0],coordenada[1],coordenada[2] 
        M += ((-1)**(fabs(i+j+k))) * (1/sqrt(i*i + j*j + k*k)) # fabs(i+j+k) porque no se admiten potencias negativas.
    
    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
    # Otra forma más eficiente que la versión 1, pero con tiempo de ejecución similar a la planteada con el bucle "for h in bucle" es la siguiente:
    # Comenta el bucle for entero y ejecuta/descomenta la siguiente línea comentada:
    # M = sum(map(lambda coor: ((-1)**(fabs(coor[0]+coor[1]+coor[2]))) * (1/sqrt(coor[0]*coor[0] + coor[1]*coor[1] + coor[2]*coor[2])), bucle))
    # - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    
    return M 

def errores(exacto: float, obtenido: float):
    # Cálculo de los errores absoluto y relativo
    Ea = fabs(exacto - obtenido)
    Er = fabs(Ea/exacto)*100
    return Ea, Er 


def main():
    inicio = default_timer()
    Mexacta = -1.74756

    for L in [20,50,100,200]:
        Mobtenida = constante_madelung(L)
        Ea, Er = errores(Mexacta, Mobtenida)
        print(f"Para L = {L} --> M = {Mobtenida}; Error absoluto = {Ea}; Error relativo = {Er}%")

    fin = default_timer()
    print()
    print(f"* Tiempo de ejecución del programa (s) = {fin - inicio}")

if __name__ == "__main__":
    main()

File: Prácticas/01-Constante Madelung/Practica1V2.cpp
#include <iostream>
#include <math.h>
#include <iomanip>
using namespace std;

#include <chrono>
using namespace chrono;

double constanteMadelung(double L){
    // Del enunciado sacamos la fórmula para M que procedemos a programar.
    // Primero inicializamos en cero la constante.
    double M = 0;

    // Tres bucles para las 3 coordenadas.
    for (int i=-L; i<=L; i++){
        for (int j=-L; j<=L; j++){
            for (int k=-L; k<=L; k++){
                if ((i==0) && (j==0) && (k==0)){
                    continue;
                } else {
                    if ((i+j+k)%2==0){
                        M = M + (1/sqrt(i*i + j*j + k*k));
                    } else {
                        M = M - (1/sqrt(i*i + j*j + k*k));
                    }
                }
            }
        }
    }

    return M;
}

void errores(double exacto, double obtenido, double &Ea, double &Er){
    // Cálculo de los errores absoluto y relativo
    Ea = fabs(exacto - obtenido);
    Er = fabs(Ea/exacto)*100;
}

int main(){
    cout<<setprecision(10);

    // Medimos el tiempo de ejecución
    time_point<system_clock> start, end;
    duration<double> elapsed_seconds;
    start = system_clock::now();

    double Mexacta = -1.74756, Mobtenida = 0., Ea = 0., Er = 0., tol = 10e-5;
    int L = 100;

    // Para L = 20
    do {
        Mobtenida = constanteMadelung(L);
        errores(Mexacta, Mobtenida, Ea, Er);
        cout<<"Para L = "<<L<<" --> M = "<<Mobtenida<<"; Error absoluto = "<<Ea<<"; Error relativo = "<<Er<<"%"<<endl;
        L = L + 100;
    } while (fabs(Ea)>tol);

    cout<<"Tolerancia alcanzada para un L = "<<L-100<<endl<<endl;

    // Terminamos de medir el tiempo de ejecución.
    end = system_clock::now();
    elapsed_seconds = end - start;
    cout<<"* Tiempo de ejecucion del programa (s) = "<<elapsed_seconds.count()<<endl;

    return 0;
}